README.md


[![Greenkeeper badge](https://badges.greenkeeper.io/carpages/gemini-respond.svg)](https://greenkeeper.io/)

CHANGELOG.md
# CHANGELOG

## v0.2.2

  - Use Carpages eslint config

## v0.2.1

  - Remove jshint in favour of eslint with custom ruleset
  - Bump travis-ci node version
  - Fix CommonJS require statement to require the correct gemini-loader module

## v0.2.0

  - Added UMD support


test/gemini-respond.edge.js
/* eslint key-spacing: ["error", { "align": "colon" }] */
requirejs.config({
  baseUrl: '../',
  paths  : {
    'underscore'    : 'bower_components/underscore/underscore',
    'jquery'        : 'bower_components/jquery/dist/jquery',
    'handlebars'    : 'bower_components/handlebars/handlebars.runtime',
    'jquery.boiler' : 'bower_components/jquery-boiler/jquery.boiler',
    'gemini.support': 'bower_components/gemini-support/gemini.support',
    'gemini'        : 'bower_components/gemini-loader/gemini'
  }
});

require([ 'gemini', 'gemini.respond' ], function( G ) {
  console.log( G.respond );
});


gruntfile.js
'use strict';

module.exports = function( grunt ) {
  // Load all grunt tasks
  require( 'load-grunt-tasks' )( grunt );

  // Show elapsed time at the end
  require( 'time-grunt' )( grunt );

  // Project configuration.
  grunt.initConfig({
    // Metadata.
    pkg: grunt.file.readJSON( 'package.json' ),
    banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' +
      '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
      '<%= pkg.homepage ? "* " + pkg.homepage + "\\n" : "" %>' +
      '* Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>;' +
      ' Licensed MIT */\n',
    // Task configuration.
    qunit: {
      all: {
        options: {
          timeout: 10000,
          urls: [ 'http://localhost:9000/test/<%= pkg.name %>.html' ],
          page: {
            viewportSize: { width: 10000, height: 10000 }
          }
        }
      }
    },
    eslint: {
      options: {
        configFile: '.eslintrc'
      },
      target: [ 'gemini.respond.js' ]
    },
    connect: {
      server: {
        options: {
          hostname: '*',
          port: 9000
        }
      }
    },
    'saucelabs-qunit': {
      all: {
        options: {
          urls: [ 'http://localhost:9000/test/<%= pkg.name %>.html' ],
          build: process.env.TRAVIS_JOB_ID,
          browsers: [
            // iOS
            {
              browserName: 'iphone',
              platform: 'OS X 10.9',
              version: '7.1'
            },
            {
              browserName: 'ipad',
              platform: 'OS X 10.9',
              version: '7.1'
            },
            // Android
            {
              browserName: 'android',
              platform: 'Linux',
              version: '4.3'
            },
            // OS X
            {
              browserName: 'safari',
              platform: 'OS X 10.9',
              version: '7'
            },
            {
              browserName: 'safari',
              platform: 'OS X 10.8',
              version: '6'
            },
            {
              browserName: 'firefox',
              platform: 'OS X 10.9',
              version: '28'
            },
            // Windows
            {
              browserName: 'internet explorer',
              platform: 'Windows 8.1',
              version: '11'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 8',
              version: '10'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '11'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '10'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '9'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '8'
            },
            {
              browserName: 'firefox',
              platform: 'Windows 7',
              version: '29'
            },
            {
              browserName: 'chrome',
              platform: 'Windows 7',
              version: '34'
            },
            // Linux
            {
              browserName: 'firefox',
              platform: 'Linux',
              version: '29'
            }
          ]
        }
      }
    }
  });

  // Default task.
  grunt.registerTask( 'default', [ 'eslint'/*, 'connect', 'qunit' */ ]);
  grunt.registerTask( 'ci', [ 'default'/*, 'saucelabs-qunit' */ ]);
};


jquery.respond.js
/**
 * @fileoverview

A jQuery plugin that allows you to listen to when the user has changed the
screen size.

### Notes
- Here's a note

### Features
- Bind events to when the screen is resized
- Only triggers the event when the resize is complete
- You have access to the current screen size according to Gemini standards
(``small``, ``medium``, ``large``, ``xlarge``)

 *
 * @namespace jquery.respond
 * @copyright Carpages.ca 2014
 * @author Matt Rose <matt@mattrose.ca>
 *
 * @requires jquery-loader
 *
 * @example
  $.respond.bind('resize', function(e, screen, windowWidth){

    console.log('Screen: ' + screen);
    console.log('Window Width: ' + windowWidth + 'px');

  });
 */

define(['jquery-loader', 'underscore'], function($, _){

  // Private settings
  var _settings = {
    delay: 500, //Delay before events are triggered
    minChange: 20 //Minimum difference in width for event to trigger
  };

  // Object to bind and trigger listeners to
  var LISTENER = $({});

  // Public object
  var plugin = {};

  // Object that specifies breakboints
  // Should sync with SASS breakpoints
  // {project}/sass/settings.scss
  // {framewrok}/variables/defaults.scss
  plugin.breakpoints = {
    'small': 0,
    'medium': 600,
    'large': 1024,
    'xlarge': 1280
  };

  /**
   * Get a list of sorted breakpoints by screen size
   *
   * @method
   * @name jquery.respond#sortedBreakpoints
   * @return {array} Array of sorted breakpoints by screen size
  **/
  plugin.sortedBreakpoints = function(){
    return _.sortBy(
      _.map(plugin.breakpoints, function(size, screen){
        return {
          size: size,
          screen: screen
        };
      }),
      function(bp){return bp.screen;}
    );
  };


  /**
   * Get the screen size based on Gemini naming conventions
   *
   * @method
   * @name jquery.respond#getScreen
   * @return {string} Screen size
  **/
  plugin.getScreen = function(){
    var returnScreen,
        width = $window.width();

    _.each(plugin.sortedBreakpoints(), function(bp){
      if(bp.size < width){
        returnScreen = bp.screen;
      }
    });

    return returnScreen;
  };

  /**
   * Check for a certain screen size or higher
   *
   * @method
   * @name jquery.respond#isScreen
   * @param {string} screen Screen size
   * @return {boolean} Whether the screen is that size or larger
  **/
  plugin.isScreen = function(screen){
    return $window.width() >= plugin.breakpoints[screen];
  };

  // Cache of the last registered width
  var _width = $window.width();

  // Add a listener to run on resize after a set delay
  $window.resize(_.debounce(function(){

    var windowWidth = $window.width();

    //Check if the window was resized enough to trigger a change
    if(Math.abs(_width - windowWidth) > _settings.minChange){
      _width = windowWidth;
      _resize(windowWidth);
    }

  }, _settings.delay));

  // Function to run when the screen is resized
  var _resize = function(windowWidth){

    //Triggers the resize event
    plugin.trigger('resize', [plugin.getScreen(), _width]);

  };

  /**
   * Bind an event
   *
   * @method
   * @name jquery.respond#bind
   * @param {string} event The event name
   * @param {function} callback The callback fuction for the event
  **/
  plugin.bind = function(){
    LISTENER.bind.apply(LISTENER, arguments);
  };
  plugin.trigger = function(){
    LISTENER.trigger.apply(LISTENER, arguments);
  };

  $.respond = plugin;

});


gemini.respond.js
/**
 * @fileoverview

A Gemini plugin that allows you to listen to when the user has changed the
screen size.

### Notes
- Here's a note

### Features
- Bind events to when the screen is resized
- Only triggers the event when the resize is complete
- You have access to the current screen size according to Gemini standards
(``small``, ``medium``, ``large``, ``xlarge``)

 *
 * @namespace gemini.respond
 * @copyright Carpages.ca 2014
 * @author Matt Rose <matt@mattrose.ca>
 *
 * @requires gemini
 *
 * @example
  G.respond.bind('resize', function(e, screen, windowWidth){

    console.log('Screen: ' + screen);
    console.log('Window Width: ' + windowWidth + 'px');

  });
 */
( function( factory ) {
  if ( typeof define === 'function' && define.amd ) {
    // AMD. Register as an anonymous module.
    define([ 'gemini' ], factory );
  } else if ( typeof exports === 'object' ) {
    // Node/CommonJS
    module.exports = factory(
      require( 'gemini-loader' )
    );
  } else {
    // Browser globals
    factory( G );
  }
}( function( $ ) {
  var _ = $._;

  // Private settings
  var _settings = {
    delay: 500, // Delay before events are triggered
    minChange: 20 // Minimum difference in width for event to trigger
  };

  // Object to bind and trigger listeners to
  var LISTENER = $({});

  // Public object
  var plugin = {};

  // Object that specifies breakboints
  // Should sync with SASS breakpoints
  // {project}/sass/settings.scss
  // {framewrok}/variables/defaults.scss
  plugin.breakpoints = {
    'small': 0,
    'medium': 600,
    'large': 1024,
    'xlarge': 1280
  };

  /**
   * Get a list of sorted breakpoints by screen size
   *
   * @method
   * @name gemini.respond#sortedBreakpoints
   * @return {array} Array of sorted breakpoints by screen size
  **/
  plugin.sortedBreakpoints = function() {
    return _.sortBy(
      _.map( plugin.breakpoints, function( size, screen ) {
        return {
          size: size,
          screen: screen
        };
      }),
      function( bp ) { return bp.size; }
    );
  };

  /**
   * Get the screen size based on Gemini naming conventions
   *
   * @method
   * @name gemini.respond#getScreen
   * @return {string} Screen size
  **/
  plugin.getScreen = function() {
    var returnScreen;
    var width = $( window ).width();

    _.each( plugin.sortedBreakpoints(), function( bp ) {
      if ( bp.size < width ) {
        returnScreen = bp.screen;
      }
    });

    return returnScreen;
  };

  /**
   * Check for a certain screen size or higher
   *
   * @method
   * @name gemini.respond#isScreen
   * @param {string} screen Screen size
   * @return {boolean} Whether the screen is that size or larger
  **/
  plugin.isScreen = function( screen ) {
    return $( window ).width() >= plugin.breakpoints[screen];
  };

  // Cache of the last registered width
  var _width = $( window ).width();

  // Add a listener to run on resize after a set delay
  $( window ).resize( _.debounce( function() {
    var windowWidth = $( window ).width();

    // Check if the window was resized enough to trigger a change
    if ( Math.abs( _width - windowWidth ) > _settings.minChange ) {
      _width = windowWidth;
      _resize( windowWidth );
    }
  }, _settings.delay ));

  // Function to run when the screen is resized
  var _resize = function( windowWidth ) {
    // Triggers the resize event
    plugin.trigger( 'resize', [ plugin.getScreen(), _width ]);
  };

  /**
   * Bind an event
   *
   * @method
   * @name gemini.respond#bind
   * @param {string} event The event name
   * @param {function} callback The callback fuction for the event
  **/
  plugin.bind = function() {
    LISTENER.bind.apply( LISTENER, arguments );
  };
  plugin.trigger = function() {
    LISTENER.trigger.apply( LISTENER, arguments );
  };

  $.respond = plugin;
}));
